packages/payment-detection/src/any/any-to-erc20-proxy-info-retriever.ts

```diff
@@ -0,0 +1,160 @@
+    private conversionProxyContractAddress: string,
+    private conversionProxyCreationBlockNumber: number,
+    private erc20FeeProxyContractAddress: string,
+    private erc20FeeProxyCreationBlockNumber: number,
+    private toAddress: string,
+    private eventName: PaymentTypes.EVENTS_NAMES,
+    private network: string,
+    private acceptedTokens?: string[],
+    private maxRateTimespan: number = 0,
+  ) {
+    // Creates a local or default provider
+    this.provider =
+      this.network === 'private'
+        ? new ethers.providers.JsonRpcProvider()
+        : ethers.getDefaultProvider(this.network);
+
+    // Setup the conversion proxy contract interface
+    this.contractConversionProxy = new ethers.Contract(
+      this.conversionProxyContractAddress,
+      erc20ConversionProxyContractAbiFragment,
+      this.provider,
+    );
+
+    this.contractERC20FeeProxy = new ethers.Contract(
+      this.erc20FeeProxyContractAddress,
+      erc20FeeProxyContractAbiFragment,
+      this.provider,
+    );
+
+    this.acceptedTokens = acceptedTokens?.map((token) => token.toLowerCase());
+  }
+
+  /**
+   * Retrieves transfer events for the current contract, address and network.
+   */
+  public async getTransferEvents(): Promise<PaymentTypes.ERC20PaymentNetworkEvent[]> {
+    // Create a filter to find all the Fee Transfer logs with the payment reference
+    const conversionFilter = this.contractConversionProxy.filters.TransferWithConversionAndReference(
+      null,
+      null,
+      '0x' + this.paymentReference,
+    ) as ethers.providers.Filter;
+    conversionFilter.fromBlock = this.conversionProxyCreationBlockNumber;
+    conversionFilter.toBlock = 'latest';
+
+    // Get the fee proxy contract event logs
+    const conversionLogs = await this.provider.getLogs(conversionFilter);
+
+    // Create a filter to find all the Fee Transfer logs with the payment reference
+    const feeFilter = this.contractERC20FeeProxy.filters.TransferWithReferenceAndFee(
+      null,
+      null,
+      null,
+      '0x' + this.paymentReference,
+      null,
+      null,
+    ) as ethers.providers.Filter;
+    feeFilter.fromBlock = this.erc20FeeProxyCreationBlockNumber;
+    feeFilter.toBlock = 'latest';
+
+    // Get the fee proxy contract event logs
+    const feeLogs = await this.provider.getLogs(feeFilter);
+
+    // Parses, filters and creates the events from the logs with the payment reference
+    const eventPromises = conversionLogs
+      // Parses the logs
+      .map((log) => {
+        const parsedLog = this.contractConversionProxy.interface.parseLog(log);
+
+        const feeLog = feeLogs.find((l) => l.transactionHash === log.transactionHash);
+        let parsedFeeLog;
+        if (feeLog) {
+          parsedFeeLog = this.contractERC20FeeProxy.interface.parseLog(feeLog);
+        }
+        return { parsedLog, log, parsedFeeLog };
+      })
+      // Keeps only the log with the right token and the right destination address
```

Comment:
If I understand properly, the reason why we cannot filter these before is for the case sensitiveness? Maybe we can clarify that ?

Reply:
And isn't it always returned as lower case by the way?

Reply:
On the v4 of ethers, It's not clear if we can filter event on with more than one value per attribute.
On the v5, it is documented. We can do this small refacto when we update ethers.

Reply:
I've added a task on JIRA